### ISSUE TITLE
Google Drive API does not always return permissionIds in files.list request

### ISSUE THREAD

**Issue by chscott** [rclone/rclone#7853]
#### The associated forum post URL from `https://forum.rclone.org`

N/A

#### What is the problem you are having with rclone?

Google Drive does not always return `permissionIds` in `files.list` request. I believe I'm seeing the same issue documented in this [StackOverflow](https://stackoverflow.com/questions/70589364/google-drive-api-does-not-return-permissionids-when-listing-files) topic.

Here's what I see in our mapper log when the problem occurs. Unfortunately, I do not have the rclone log, but this logging just prints the raw object received from rclone:

```
[1mZ2lGOWiTo1fZ9mToOwy7MoELJlhIlH_] Object from rclone: {"SrcFs":"Source{xrgXa}:","SrcFsType":"drive","DstFs":"Target{qpiyZ}:","DstFsType":"onedrive","Remote":"REDACTED.pdf","Size":97263,"MimeType":"application/pdf","ModTime":"2024-01-17T20:29:42.252Z","IsDir":false,"ID":"1mZ2lGOWiTo1fZ9mToOwy7MoELJlhIlH_","Metadata":{"btime":"2024-01-17T20:30:45Z","content-type":"application/pdf","copy-requires-writer-permission":"false","mtime":"2024-01-17T20:29:42.252Z","starred":"false","viewed-by-me":"true","writers-can-share":"false"}}
```

Note that there is no `permissions` property.

We enabled verbose logging for a second attempt, but the file wasn't deleted in the target, so it wasn't marked for transfer. But I see this in the rclone log:

```
{
  "copyRequiresWriterPermission": false,
  "md5Checksum": "a11510a3be225916b8e61fe37d0bd81a",
  "viewedByMe": true,
  "mimeType": "application/pdf",
  "parents": ["1WLFh3GPi4yuo14my91XXdhKEl3LVLkwr"],
  "webViewLink": "https://drive.google.com/file/d/1mZ2lGOWiTo1fZ9mToOwy7MoELJlhIlH_/view?usp=drivesdk",
  "size": "97263",
  "id": "1mZ2lGOWiTo1fZ9mToOwy7MoELJlhIlH_",
  "name": "REDACTED.pdf",
  "starred": false,
  "trashed": false,
  "explicitlyTrashed": false,
  "createdTime": "2024-01-17T20:30:45.000Z",
  "modifiedTime": "2024-01-17T20:29:42.252Z",
  "viewedByMeTime": "2024-01-17T20:30:16.047Z",
  "hasAugmentedPermissions": false,
  "permissionIds": ["08143305669083284370", "09148056542529227285", "05730075870274744324", "14428220901927124198", "10333761160097088327", "01188868095341529512", "11263538924981798501", "15302165612694109360", "15968751374380107705", "01848923844518485244", "17043712991773318823", "00993337219959399960", "13488475314984407437", "15844604519322965153", "08826457509796904545", "12361398014948882636", "14049068988452217961", "17543280361655486121", "16152041014387613021", "14207865859295967165", "02698292558398922638", "08631903179092097718", "06109885187853851371", "04455832434476281173", "12302732694699397834", "07136569753069756070", "07122999719496272434", "17295166189381661175", "02012353872150768475", "13316925832582353280", "03558791494311673145", "06079240731634784000", "01693074727832241524", "16323005513084617832", "09884135463593788567", "05699788620167125482", "07088479073237879776", "11978124046481694590", "02068273953380346298"]
}
```

When metadata is enabled and a given file does not have the `permissionIds` property set, can we issue a request to get that information for the file in question, thereby working around this apparent API defect?

#### What is your rclone version (output from `rclone version`)

```
rclone v1.67.0-beta.7957.625622bcb.fix-onedrive-metadata
- os/version: Microsoft Windows 11 Pro 23H2 (64 bit)
- os/kernel: 10.0.22631.3527 (x86_64)
- os/type: windows
- os/arch: amd64
- go/version: go1.22.3
- go/linking: static
- go/tags: cmount
```

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)

Windows 11, 64 bit

#### Which cloud storage system are you using? (e.g. Google Drive)

Google Drive

#### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)

`rclone copy Source: Target:`

#### A log from the command with the `-vv` flag (e.g. output from `rclone -vv copy /tmp remote:tmp`)

I'll email these separately. Unfortunately, this issue is intermittent, and we saw the issue before enabling verbose logging.

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by ncw** [comment]
Heads up @rclone/support - the "Support Contract" label was applied to this issue.

**Comment by ncw** [comment]
We only add `permissionIds` if we didn't have any `permissions`.

https://github.com/rclone/rclone/blob/7b89735ae7263cf24669a518a5508f2782235036/backend/drive/metadata.go#L379-L383

Perhaps this is the problem? I think the reasoning behind this is that decoded permissions will have everything you need in so only show this if not present. This could be a bit cleverer and only show permisisonIds that aren't in permissions.

---- 

From the stack overflow post this looks like a bug in google drive (rclone already has quite a few workarounds for them).

If you do an `rclone -M -R --disable ListR lsjson source:` on the drive remote then this should be identical to what rclone is doing when it is doing an `rclone -M sync/copy`

Can you do that a few times and see if you can see flakiness?

I suggest you pipe it through `jq` to make it easy to read, then `diff` the trials.

I didn't see a bug in the google bug tracker about this.

> When metadata is enabled and a given file does not have the permissionIds property set, can we issue a request to get that information for the file in question, thereby working around this apparent API defect?

Possibly but depending on exactly what is happening we may not be able to tell the permissionsIds are missing or not.

**Comment by chscott** [comment]
@ncw The HTTP response from the test shows the presence of the `permissionIds` property, but the `lsjson` output does not have any permissions-related information in the metadata.

**HTTP Response**

```
{
  "copyRequiresWriterPermission": false,
  "md5Checksum": "a11510a3be225916b8e61fe37d0bd81a",
  "viewedByMe": true,
  "mimeType": "application/pdf",
  "parents": ["1WLFh3GPi4yuo14my91XXdhKEl3LVLkwr"],
  "webViewLink": "https://drive.google.com/file/d/1mZ2lGOWiTo1fZ9mToOwy7MoELJlhIlH_/view?usp=drivesdk",
  "size": "97263",
  "id": "1mZ2lGOWiTo1fZ9mToOwy7MoELJlhIlH_",
  "name": "REDACTED.pdf",
  "starred": false,
  "trashed": false,
  "explicitlyTrashed": false,
  "createdTime": "2024-01-17T20:30:45.000Z",
  "modifiedTime": "2024-01-17T20:29:42.252Z",
  "viewedByMeTime": "2024-01-17T20:30:16.047Z",
  "hasAugmentedPermissions": false,
  "permissionIds": ["08143305669083284370", "09148056542529227285", "05730075870274744324", "14428220901927124198", "10333761160097088327", "01188868095341529512", "11263538924981798501", "15302165612694109360", "15968751374380107705", "01848923844518485244", "17043712991773318823", "00993337219959399960", "13488475314984407437", "15844604519322965153", "08826457509796904545", "12361398014948882636", "14049068988452217961", "17543280361655486121", "16152041014387613021", "14207865859295967165", "02698292558398922638", "08631903179092097718", "06109885187853851371", "04455832434476281173", "12302732694699397834", "07136569753069756070", "07122999719496272434", "17295166189381661175", "02012353872150768475", "13316925832582353280", "03558791494311673145", "06079240731634784000", "01693074727832241524", "16323005513084617832", "09884135463593788567", "05699788620167125482", "07088479073237879776", "11978124046481694590", "02068273953380346298"]
}
```

**`lsjson` output**

```
[{
        "Path": "REDACTED.pdf",
        "Name": "REDACTED.pdf",
        "Size": 97263,
        "MimeType": "application/pdf",
        "ModTime": "2024-01-17T20:29:42.252Z",
        "IsDir": false,
        "Hashes": {
            "md5": "a11510a3be225916b8e61fe37d0bd81a"
        },
        "ID": "1mZ2lGOWiTo1fZ9mToOwy7MoELJlhIlH_",
        "Metadata": {
            "btime": "2024-01-17T20:30:45.000Z",
            "content-type": "application/pdf",
            "copy-requires-writer-permission": "false",
            "mtime": "2024-01-17T20:29:42.252Z",
            "starred": "false",
            "viewed-by-me": "true",
            "writers-can-share": "false"
        }
    }
]
```

**Comment by chscott** [comment]
From a data perspective, it looks like this is another case of permissions not being propagated because they are not set directly on the file (i.e., `"hasAugmentedPermissions": false). So I think things are being filtered out here:

https://github.com/rclone/rclone/blob/7b89735ae7263cf24669a518a5508f2782235036/backend/drive/metadata.go#L373-L377

These being inherited permissions was not my understanding when the issue was raised to me, so I'll pursue it on my end for now. I think it would be helpful to log something (perhaps at the DEBUG level) in this spot so it is clear why permissions are not being propagated.

**Comment by chscott** [comment]
@ncw The debug statement works well. Thanks! I think we can consider this issue resolved.

**Comment by ncw** [comment]
Great!

I've merged the debug to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.67

I'll close this for now.
